### PR TITLE
Refactor lw

### PR DESCRIPTION
**Pull request by ros3bud**
@cgrayson, these changes reflect the changes we discussed when I created the personal loan integration, to dry up the different modules. Specifically, the code in parse.coffee was moved to common.coffee. I also moved the response testing into the common_spec file. LGTY?

### CODE REVIEW

File: spec/common_spec.coffee [hunk 1]
```diff
@@ -280,7 +282,54 @@ describe 'Employment Status', ->
...
+                 <ERRSTATUS>
+                   <ERRNUM>0</ERRNUM>
+                   <ERRDESCRIPTION>Success</ERRDESCRIPTION>
+                 </ERRSTATUS>
+               </LTACK>
+               '''
+
+    for form in formsArray()
+      do (form) ->
+        res = form.response(req, outcome: 'success')
```
Comment (cgrayson) [inline spec/common_spec.coffee]: This is interesting, but not what I had in mind. Since this is common code, it can just be straight tested as such. That is, not keeping and iterating over an index of all the users of this code, but just testing `common.response()`.

File: spec/personalloan_spec.coffee [hunk 2]
```diff
@@ -28,20 +28,14 @@ describe 'Personal loan', ->
...
       it 'should parse phone_1 when there is an area code and extension', ->
         assert.equal personalloan.request(req(xml({contactPhone: 5555555, areaCode: 512, extension: 1234}))).phone_1, '5125555555 x1234'
 
-      it 'should parse a masked lead', ->
-        assert.equal personalloan.request(req(maskedXml({contactPhone: 5555555, areaCode: 512, extension: 789}))).phone_1, '512******* x789'
-
-      it 'should parse a masked lead with no area code or extension', ->
-        assert.equal personalloan.request(req(maskedXml(contactPhone: 5555555))).phone_1, '*******'      
+      it 'should parse phone_1 when it is an 800 number with a pin', ->
+        assert.equal personalloan.request(req(xml({contactPhone: 5551212, areaCode: 800, extension: 789}))).phone_1, '8005551212 x789'
```
Comment (cgrayson) [inline spec/personalloan_spec.coffee]: I don't follow what's up with these 3 deleted and 1 added test; are these changes related to the refactoring?

File: src/common.coffee [hunk 3]
```diff
@@ -40,6 +40,212 @@ response.variables = ->
...
+      when 'carfinancing' then result.push 'car finance'
+      when 'payoffheloan' then result.push 'pay off home equity loan'
+      when 'payoffheloc' then result.push 'pay off home equity line of credit'
+      when 'payoffsecond' then result.push 'pay off second mortgage'
+      when 'greenloan' then result.push 'green loan'
+      else result.push reason
+
+  result
+
+parseTrueOrFalse = (value) ->
```
Comment (cgrayson) [inline src/common.coffee]: You know what, I think this function predates the existence of our `boolean` type. For `boolean` request variables, I don't think this is necessary anymore.

File: src/common.coffee [hunk 4]
```diff
@@ -40,6 +40,212 @@ response.variables = ->
...
+  return null unless value?.trim()
+  value = value.toLowerCase()
+  switch value
+    when 'y', 'yes', 'true', 't' then true
+    when 'n', 'no', 'false', 'f' then false
+    else null
+
+parseDate = (value, format='MM/DD/YYYY') ->
```
Comment (cgrayson) [inline src/common.coffee]: Similar to `parseTrueOrFalse()`, I don't think this is needed for `date` types.

File: src/common.coffee [hunk 5]
```diff
@@ -40,6 +40,212 @@ response.variables = ->
...
+    when 'fulltime' then 'full time'
+    when 'parttime' then 'part time'
+    when 'selfemployed' then 'self employed'
+    when 'homemaker', 'student', 'retired', 'unemployed', 'other' then value
+    else
+      null
+
+#not used in unified
+
+parseTime = (value) ->
```
Comment (cgrayson) [inline src/common.coffee]: Similar to `parseTrueOrFalse()`, don't think this is needed for `time` types.

File: src/common.coffee [hunk 6]
```diff
@@ -40,6 +40,212 @@ response.variables = ->
...
+  value = value.toLowerCase()
+  switch value
+    when 'fulltime' then 'full time'
+    when 'parttime' then 'part time'
+    when 'selfemployed' then 'self employed'
+    when 'homemaker', 'student', 'retired', 'unemployed', 'other' then value
+    else
+      null
+
+#not used in unified
```
Comment (cgrayson) [inline src/common.coffee]: Realizing this comment was just brought over from `parse.coffee`, I don't know how informative it is.

File: src/common.coffee (hunk 3 above)
Comment (ros3bud) [inline src/common.coffee]: Ok, seems like we should add an issue for this.

File: src/common.coffee (hunk 4 above)
Comment (ros3bud) [inline src/common.coffee]: And this.

File: src/common.coffee (hunk 5 above)
Comment (ros3bud) [inline src/common.coffee]: And this.

File: spec/personalloan_spec.coffee (hunk 2 above)
Comment (ros3bud) [inline spec/personalloan_spec.coffee]: They're not and I should have mentioned that. I can add an issue.

**Comment by cgrayson** [comment]
This is coming along, and will be a nice cleanup.

File: spec/common_spec.coffee (hunk 1 above)
Comment (ros3bud) [inline spec/common_spec.coffee]: Ok, I think I see what you mean.

**Comment by ros3bud** [comment]
Awesome, thanks :)

File: src/common.coffee (hunk 6 above)
Comment (ros3bud) [inline src/common.coffee]: Ok, yeah, I see what you mean.

**Comment by ros3bud** [comment]
Back to you, @cgrayson - just let me know if you have any questions.

File: spec/common_spec.coffee [hunk 7]
```diff
@@ -141,27 +141,27 @@ describe 'Loan Reason', ->
...
+    assert.equal request('mortgage.application.is_home_found': 'N'), 'N'
+    assert.equal request('mortgage.application.is_home_found': 'no'), 'no'
+    assert.equal request('mortgage.application.is_home_found': 'false'), 'false'
+    assert.equal request('mortgage.application.is_home_found': 'f'), 'f'
 
   it 'should parse null', ->
-    assert.equal request('mortgage.application.is_home_found': 'nope'), null
-    assert.equal request('mortgage.application.is_home_found': ' '), null
+    assert.equal request('mortgage.application.is_home_found': 'nope'), 'nope'
+    assert.equal request('mortgage.application.is_home_found': ' '), ' '
```
Comment (cgrayson) [inline spec/common_spec.coffee]: I don't think these three `it()`s need to be as extensive anymore, now that the integration is relying on common (and presumably well-tested) code from the `leadconduit-types` module. If anything, a test that the input value isn't altered would be enough.

File: spec/common_spec.coffee [hunk 8]
```diff
@@ -2,25 +2,25 @@ assert    = require('chai').assert
...
-    assert.equal request(credit_rating: 'GOOD'), 'good'
+    assert.equal request(credit_rating: 'GOOD'), 'GOOD'
 
   it 'should parse fair credit', ->
-    assert.equal request(credit_rating: 'FAIR'), 'fair'
+    assert.equal request(credit_rating: 'FAIR'), 'FAIR'
 
   it 'should parse poor credit', ->
-    assert.equal request(credit_rating: 'POOR'), 'poor'
+    assert.equal request(credit_rating: 'POOR'), 'POOR'
```
Comment (cgrayson) [inline spec/common_spec.coffee]: I think I commented on a commit instead of here on the main PR line, so forgive my repetition, but: I'm not sure why you changed this behavior to no longer lowercase these values? This could be a breaking change to existing flows.

File: spec/common_spec.coffee [hunk 9]
```diff
@@ -229,9 +229,9 @@ describe 'Discharged Months', ->
 describe 'Time to Contact', ->
 
   it 'should parse anytime', ->
-    assert.equal request('time_to_contact': '0000'), 'anytime'
-    assert.equal request('time_to_contact': 'something unexpected'), 'anytime'
-    assert.equal request('time_to_contact': ' '), 'anytime'
+    assert.equal request('time_to_contact': '0000'), 'any time'
+    assert.equal request('time_to_contact': 'something unexpected'), 'something unexpected'
+    assert.equal request('time_to_contact': ' '), ' '
```
Comment (cgrayson) [inline spec/common_spec.coffee]: Similar to my question about credit rating, this change could affect existing flows that depend on matching to `anytime`.

File: spec/shortform_spec.coffee [hunk 10]
```diff
@@ -78,72 +78,72 @@ describe 'Shortform', ->
...
 
       it 'should parse 5 years since bankruptcy', ->
         body = xml(Bankruptcy: '5-7 years ago')
-        assert.isTrue shortform.request(req(body)).mortgage.bankruptcy.was_declared
+        assert.equal shortform.request(req(body)).mortgage.bankruptcy.was_declared, true
 
       it 'should parse never declared bankruptcy', ->
         body = xml(Bankruptcy: 'Never declared bankruptcy')
-        assert.isFalse shortform.request(req(body)).mortgage.bankruptcy.was_declared
+        assert.equal shortform.request(req(body)).mortgage.bankruptcy.was_declared, false
```
Comment (cgrayson) [inline spec/shortform_spec.coffee]: I don't think these 6 changes are needed (from `.isTrue()` to `.isEqual(…, true)`)

File: src/longform.coffee [hunk 11]
```diff
@@ -45,24 +44,24 @@ request = (req) ->
...
     application:
       # timeframe_days: parseTimeframeDays(lead.TimeLine) # could use EstCloseDate?
-      created_at: parse.time(lead.Application.LoanRequestCompleteDate)
+      created_at: lead.Application.LoanRequestCompleteDate
       credit_score: lead.Borrower.LenderAssignedCreditValue
-      bankruptcy: parse.bankruptcy(lead.Borrower.Declaration.DeclaredBankruptcy)
-      working_with_realtor: parse.trueOrFalse(lead.TransactionType.Mortgage.HaveRealtor) # deprecated
-      is_working_with_realtor: parse.trueOrFalse(lead.TransactionType.Mortgage.HaveRealtor)
+      bankruptcy: common.bankruptcy(lead.Borrower.Declaration.DeclaredBankruptcy)
+      working_with_realtor: lead.TransactionType.Mortgage.HaveRealtor
```
Comment (cgrayson) [inline src/longform.coffee]: Minor, but I'd keep the comment that `working_with_realtor` being a deprecated field.

File: src/shortform.coffee [hunk 12]
```diff
@@ -20,23 +19,23 @@ parseLead = (xml) ->
...
     first_mortgage_balance: lead.FirstMortgageBalance ? 0
 
     application:
-      timeframe_days: parse.timeframeDays(lead.TimeLine)
-      created_at: parse.date(lead.ApplicationDate)
-      credit_rating: parse.creditRating(lead.CreditRating) # deprecated
-      bankruptcy: parse.bankruptcy(lead.Bankruptcy) #deprecated
-      working_with_realtor: parse.trueOrFalse(lead.WorkingWithRealtor) # deprecated
-      is_working_with_realtor: parse.trueOrFalse(lead.WorkingWithRealtor)
-      is_military: parse.trueOrFalse(lead.IsMilitary)
```
Comment (cgrayson) [inline src/shortform.coffee]: Minor: as elsewhere, a couple more `# deprecated` comments lost (and one kept…? :wink: ) here.

File: spec/common_spec.coffee (hunk 8 above)
Comment (ros3bud) [inline spec/common_spec.coffee]: I changed it because I thought that the outbound integration should handle making the data consistent in this way. But based on our recent conversation, I see that is not needed.

File: spec/common_spec.coffee (hunk 9 above)
Comment (ros3bud) [inline spec/common_spec.coffee]: It seems like we should have official values which would go in the time_to_contact field and make a decision about whether this should always be "anytime" or "any time". For example, when I built the personal loan inbound integration, we decided that "any time" would be the best fit here.

File: spec/shortform_spec.coffee (hunk 10 above)
Comment (ros3bud) [inline spec/shortform_spec.coffee]: I did this because I noticed that when a value with isFalse wrong, I don't get the line number with the origin of the error back. In addition, because I changed some of them it made sense to me to be consistent.

File: src/longform.coffee (hunk 11 above)
Comment (ros3bud) [inline src/longform.coffee]: Ok, makes sense.

File: src/shortform.coffee (hunk 12 above)
Comment (ros3bud) [inline src/shortform.coffee]: Ok - np.

File: spec/common_spec.coffee (hunk 7 above)
Comment (ros3bud) [inline spec/common_spec.coffee]: Ok, I think I see what you mean.

File: spec/common_spec.coffee (hunk 9 above)
Comment (cgrayson) [inline spec/common_spec.coffee]: To follow up on this particular difference, 'anytime' is the way to go, since it's been in use longer (I don't see any production flows using the `personal_loan` inbound yet).

File: spec/shortform_spec.coffee (hunk 10 above)
Comment (cgrayson) [inline spec/shortform_spec.coffee]: Hmm, that's odd. I don't see much difference in how line numbers are reported when I run `npm test`; maybe you're running them another way? 

```
  1) Shortform Request Application should parse 0 years since bankruptcy:
     AssertionError: expected true to equal false
    at Context.<anonymous> (spec/shortform_spec.coffee:107:16)
```
vs. 

```
  1) Shortform Request Application should parse 0 years since bankruptcy:
     AssertionError: expected true to be false
    at Function.assert.isFalse (node_modules/chai/lib/chai/interface/assert.js:353:31)
    at Context.<anonymous> (spec/shortform_spec.coffee:107:16)
```

File: spec/common_spec.coffee (hunk 9 above)
Comment (ros3bud) [inline spec/common_spec.coffee]: Ok, sounds good by me.

File: spec/shortform_spec.coffee (hunk 10 above)
Comment (ros3bud) [inline spec/shortform_spec.coffee]: Ok, I can reverse it back, but I think we are running things differently. For example, this is what I get when I change an expected boolean: 

  1) Shortform Request Application should parse 5 years since bankruptcy:
     AssertionError: expected true to be false

**Comment by ros3bud** [comment]
Back to you, @cgrayson - questions welcome.

File: spec/shortform_spec.coffee [hunk 13]
```diff
@@ -78,28 +78,28 @@ describe 'Shortform', ->
...
 
       it 'should parse fair credit rating', ->
         body = xml(CreditRating: 'FAIR')
-        assert.equal shortform.request(req(body)).credit_rating, 'fair'
+        assert.equal shortform.request(req(body)).credit_rating, 'FAIR'
 
       it 'should parse good credit rating', ->
         body = xml(CreditRating: 'POOR')
-        assert.equal shortform.request(req(body)).credit_rating, 'poor'
+        assert.equal shortform.request(req(body)).credit_rating, 'POOR'
```
Comment (cgrayson) [inline spec/shortform_spec.coffee]: These 5 changes are still changing existing behavior.

File: spec/common_spec.coffee (hunk 9 above)
Comment (cgrayson) [inline spec/common_spec.coffee]: These changes should be reverted, right?

File: spec/common_spec.coffee [hunk 14]
```diff
@@ -141,27 +141,26 @@ describe 'Loan Reason', ->
...
+    assert.equal request('mortgage.application.is_home_found': 'N'), 'N'
+    assert.equal request('mortgage.application.is_home_found': 'no'), 'no'
+    assert.equal request('mortgage.application.is_home_found': 'false'), 'false'
+    assert.equal request('mortgage.application.is_home_found': 'f'), 'f'
 
-  it 'should parse null', ->
-    assert.equal request('mortgage.application.is_home_found': 'nope'), null
-    assert.equal request('mortgage.application.is_home_found': ' '), null
+  it 'should parse a string', ->
+    assert.equal request('mortgage.application.is_home_found': 'nope'), 'nope'
```
Comment (cgrayson) [inline spec/common_spec.coffee]: As before, these tests are no longer needed.

File: spec/common_spec.coffee [hunk 15]
```diff
@@ -18,9 +18,9 @@ describe 'Credit Rating', ->
     assert.equal request(credit_rating: 'POOR'), 'poor'
 
   it 'should parse null credit', ->
-    assert.equal request(credit_rating: 'WHOOPS'), null
+    assert.equal request(credit_rating: 'WHOOPS'), 'whoops'
     assert.equal request(credit_rating: ' '), null
-    assert.equal request(credit_rating: 37), null
+    assert.equal request(credit_rating: 37), '37'
```
Comment (cgrayson) [inline spec/common_spec.coffee]: These change existing behavior.

File: spec/common_spec.coffee (hunk 9 above)
Comment (ros3bud) [inline spec/common_spec.coffee]: Yes, reverted to use 'any time'.

File: spec/common_spec.coffee (hunk 14 above)
Comment (ros3bud) [inline spec/common_spec.coffee]: Ok.

File: src/common.coffee [hunk 16]
```diff
@@ -2,8 +2,15 @@ mimecontent = require('mime-content')
 xmlbuilder = require('xmlbuilder')
 crypto = require('crypto')
 helper = require('./helper')
-parse = require('./parse')
+moment = require('moment')
```
Comment (cgrayson) [inline src/common.coffee]: Doesn't look like `moment` is needed here anymore (or `mimecontent` or `helper`, actually).

File: src/longform.coffee [hunk 17]
```diff
@@ -45,24 +44,24 @@ request = (req) ->
...
     application:
       # timeframe_days: parseTimeframeDays(lead.TimeLine) # could use EstCloseDate?
-      created_at: parse.time(lead.Application.LoanRequestCompleteDate)
+      created_at: lead.Application.LoanRequestCompleteDate
       credit_score: lead.Borrower.LenderAssignedCreditValue
-      bankruptcy: parse.bankruptcy(lead.Borrower.Declaration.DeclaredBankruptcy)
-      working_with_realtor: parse.trueOrFalse(lead.TransactionType.Mortgage.HaveRealtor) # deprecated
-      is_working_with_realtor: parse.trueOrFalse(lead.TransactionType.Mortgage.HaveRealtor)
+      bankruptcy: common.bankruptcy(lead.Borrower.Declaration.DeclaredBankruptcy) #deprecated
+      working_with_realtor: lead.TransactionType.Mortgage.HaveRealtor
```
Comment (cgrayson) [inline src/longform.coffee]: It seems that the `# deprecated` comment has moved from `working_with_realtor` to `bankruptcy`. While the latter [is indeed deprecated in `fields`](https://github.com/activeprospect/leadconduit-fields/blob/master/fields/mortgage.json#L84), that hasn't been applied to this integration yet.

File: spec/common_spec.coffee [hunk 18]
```diff
@@ -141,27 +141,23 @@ describe 'Loan Reason', ->
...
-    assert.equal request('mortgage.application.is_home_found': 'false'), false
-    assert.equal request('mortgage.application.is_home_found': 'f'), false
-
-  it 'should parse null', ->
-    assert.equal request('mortgage.application.is_home_found': 'nope'), null
-    assert.equal request('mortgage.application.is_home_found': ' '), null
+    assert.equal request('mortgage.application.is_home_found': 'N'), 'N'
+    assert.equal request('mortgage.application.is_home_found': 'no'), 'no'
+    assert.equal request('mortgage.application.is_home_found': 'false'), 'false'
+    assert.equal request('mortgage.application.is_home_found': 'f'), 'f'
```
Comment (cgrayson) [inline spec/common_spec.coffee]: Before, when this integration had its own custom boolean-parsing function (`parseTrueOrFalse()`), it was good to test a variety of true, false, and invalid values. But now that it's using the common (and presumably well-tested) code from the `leadconduit-types` module to do that parsing work, it's no longer necessary to test all of these values here. 

I [said before](https://github.com/activeprospect/leadconduit-integration-lendingtree/pull/33#discussion-diff-50148177), "If anything, a test that the input value isn't altered would be enough". What I meant is, you could keep _one_ of these tests, and it would verify that your code will set `mortgage.application.is_home_found` to the right value when the input XML has:

```xml
  <ConsumerProfileInformation>
    <ProductProfileInformation>
      <FoundHome>Y</FoundHome>
    …
```

File: spec/common_spec.coffee [hunk 19]
```diff
@@ -230,8 +226,8 @@ describe 'Time to Contact', ->
 
   it 'should parse anytime', ->
     assert.equal request('time_to_contact': '0000'), 'anytime'
-    assert.equal request('time_to_contact': 'something unexpected'), 'anytime'
-    assert.equal request('time_to_contact': ' '), 'anytime'
+    assert.equal request('time_to_contact': 'something unexpected'), 'something unexpected'
+    assert.equal request('time_to_contact': ' '), ' '
```
Comment (cgrayson) [inline spec/common_spec.coffee]: These changes alter existing behavior and may cause problems in production flows.

File: src/personalloan.coffee [hunk 20]
```diff
@@ -28,14 +27,14 @@ request = (req) ->
   city: consumerInfo.ContactCity
   state: consumerInfo.ContactState
   postal_code: consumerInfo.ContactZip
-  time_to_contact: parseTimeToContact(consumerInfo.TimeToContact)
-  email: (consumerInfo.EmailAddress._ unless consumerInfo.EmailAddress.IsMasked is 'Y') ? '**********'
+  time_to_contact: common.timeToContact(consumerInfo.TimeToContact)
+  email: consumerInfo.EmailAddress._
   phone_1: getPhone1(consumerInfo.ConsumerGeoPhoneAreaCode, consumerInfo.ContactPhone._, consumerInfo.ContactPhone.IsMasked, consumerInfo.ContactPhoneExtension)
-  dob: consumerProfileInfo.DateOfBirth
+  dob: consumerProfileInfo.DateOfBirth.valueOf()
```
Comment (cgrayson) [inline src/personalloan.coffee]: I don't think you want to `valueOf()` here; it's a `Date` object and appending it to the lead as such is fine. (The test may `valueOf()` it, for comparison, though).

File: spec/common_spec.coffee (hunk 19 above)
Comment (ros3bud) [inline spec/common_spec.coffee]: Ok. This is different from other refactorings of old code that I've done, but I've changed the parseTimeToContact method back to return null when there is no match.

File: src/personalloan.coffee (hunk 20 above)
Comment (ros3bud) [inline src/personalloan.coffee]: Ok, looks like I don't need valueOf() at all.

File: src/common.coffee [hunk 21]
```diff
@@ -1,10 +1,16 @@
 mimecontent = require('mime-content')
```
Comment (cgrayson) [inline src/common.coffee]: Repeating on the PR instead of the commit (sorry about that): As noted in the last PR, `mimecontent` is not needed here.

File: spec/common_spec.coffee [hunk 22]
```diff
@@ -230,8 +220,8 @@ describe 'Time to Contact', ->
 
   it 'should parse anytime', ->
     assert.equal request('time_to_contact': '0000'), 'anytime'
-    assert.equal request('time_to_contact': 'something unexpected'), 'anytime'
-    assert.equal request('time_to_contact': ' '), 'anytime'
+    assert.equal request('time_to_contact': 'something unexpected'), null
+    assert.equal request('time_to_contact': ' '), null
```
Comment (cgrayson) [inline spec/common_spec.coffee]: Repeating on PR vs. commit: As noted previously, these changes represent a change to existing behavior and may cause problems in production flows.

Please let me know if it's unclear why this is a problem (or let me know what I'm missing here).

File: src/common.coffee (hunk 21 above)
Comment (ros3bud) [inline src/common.coffee]: Yes. My fault! I thought I had taken it out.

File: spec/common_spec.coffee (hunk 22 above)
Comment (ros3bud) [inline spec/common_spec.coffee]: I will ping you.

**Comment by cgrayson** [comment]
:+1: :+1: :exclamation:
